Mutect2 multi sample subworkflow

berguner opened the pull request:
<!--
# nf-core/sarek pull request

Many thanks for contributing to nf-core/sarek!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the dev branch, unless you're preparing a pipeline release.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/nf-core/sarek/tree/master/.github/CONTRIBUTING.md)
-->
~~Create a subworkflow~~ run mutect2 in joint mode for patients with multiple tumor samples as described in #916 .
~~- Created 2 subworkflows `BAM_VARIANT_CALLING_TUMOR_ONLY_MUTECT2_MS` and `BAM_VARIANT_CALLING_SOMATIC_MUTECT2_MS`.~~
- Added the ~~`--mutect2_multi_sample`~~  `--joint_mutect` pipeline parameter to switch on multi sample mode.
- Added test data for multi sample mode
~~- VCFs are saved under `variant_calling/mutect2_ms` in multi sample mode.~~

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] Make sure your code lints (`nf-core lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [x] `CHANGELOG.md` is updated.
- [x] `README.md` is updated (including new tool citations and authors/contributors).

maxulysse commented:
Hi @berguner sorry for the delay.
I'm afraid we have some merge conflicts to solve there.
You seem to have used slightly out of date code on you PR.
Are you ok into trying to solve these conflicts, or do you need any help with that?

maxulysse commented on nextflow.config:
```diff
@@ -176,6 +177,7 @@ profiles {
         charliecloud.enabled   = false
         conda.enabled          = false
         docker.enabled         = true
+        docker.registry        = 'quay.io'
```
```suggestion
```
This is done in line 297/300

maxulysse commented:
you have tiny issues with end of lines in some files, but this is looking super good.

berguner commented:
I just run `nf-core lint --fix files_unchanged`, I don't know how else I can sort out the linting issues. Should I run prettier separately?

maxulysse commented:
errors are `Wrong line endings or no final newline`, so it's not `prettier`, but similar, let me check what can be done.
Maybe a @nf-core-bot fix linting pretty please will do the trick

maxulysse commented:
ok, so some fixes are done, but not all.
Do you mind if I push commit in your branch?

berguner commented:
Sure, I just added you as a collaborator to my repo.

maxulysse commented:
Thanks a lot, we should be ok with the linting now

maxulysse commented:
Do you mind if I split the test samplesheet in tumor_only and regular somatic (tumor/normal)?
Adding some basic manual tests while I'm at it.

Plus, do you mind if I rename all `tumour` to `tumor` to be consistent?

maxulysse commented:
Sorry about the incoming changes then ;-)

maxulysse commented on nextflow_schema.json:
```diff
@@ -271,7 +276,7 @@
                     "type": "number",
                     "fa_icon": "fas fa-bacon",
-                    "help_text": "ASCAT: optional argument to override ASCAT optimization and supply psi parameter (expert parameter, don\u2019t adapt unless you know what you\u2019re doing). See [here](https://raw.githubusercontent.com/VanLoo-lab/ascat/master/man/ASCAT-manual.pdf)",
+                    "help_text": "ASCAT: optional argument to override ASCAT optimization and supply psi parameter (expert parameter, don’t adapt unless you know what you’re doing). See [here](https://raw.githubusercontent.com/VanLoo-lab/ascat/master/man/ASCAT-manual.pdf)",
```
```suggestion
                    "help_text": "ASCAT: optional argument to override ASCAT optimization and supply psi parameter (expert parameter, do not adapt unless you know what you are doing). See [here](https://raw.githubusercontent.com/VanLoo-lab/ascat/master/man/ASCAT-manual.pdf)",
```

maxulysse commented:
@berguner what do you think about changing the output folder from `mutect2_ms/ to `mutect2/multi_sample`.
Also not sure about the name for the variant caller, I don't think it should be `mutect2_ms`, `mutect2` doesn't seem to cover all, any idea?

maxulysse commented:
tests are coming, just need to finish them locally

FriederikeHanssen commented on conf/modules/mutect2.config:
```diff
@@ -14,110 +14,197 @@
+    else if (params.tools && params.tools.split(',').contains('mutect2') && params.mutect2_multi_sample) {
+        // MULTI SAMPLE MUTECT2
+        withName: 'GATK4_MUTECT2' {
+            ext.prefix       = { meta.num_intervals <= 1 ? "${meta.id}.mutect2" : "${meta.id}.mutect2.${intervals.simpleName}" }
+            ext.when         = { params.tools && params.tools.split(',').contains('mutect2') && params.mutect2_multi_sample }
+            ext.args         = { params.ignore_soft_clipped_bases ? "--dont-use-soft-clipped-bases true --f1r2-tar-gz ${task.ext.prefix}.f1r2.tar.gz" : "--f1r2-tar-gz ${task.ext.prefix}.f1r2.tar.gz" }
```
can we generalize this one for both mutect2 usages?

FriederikeHanssen commented on subworkflows/local/bam_variant_calling_somatic_mutect2_ms/main.nf:
```diff
@@ -0,0 +1,324 @@
+    // Separate tumor cram files
+    ch_tumor_cram = input_intervals.map{meta, n_cram, n_crai, t_cram, t_crai, intervals -> [[id: meta.patient, patient: meta.patient, normal_id: meta.normal_id, sex: meta.sex, num_intervals: meta.num_intervals], t_cram, t_crai]}.unique()
+    // Merge normal and tumor samples by patient
+    ch_tn_cram = ch_normal_cram_first.mix(ch_tumor_cram).groupTuple()
```
`size/groupKey`

berguner replied:
See https://github.com/nf-core/sarek/pull/1013#discussion_r1241810307

FriederikeHanssen commented on subworkflows/local/bam_variant_calling_somatic_mutect2_ms/main.nf:
```diff
@@ -0,0 +1,324 @@
+    // Merge normal and tumor samples by patient
+    ch_tn_cram = ch_normal_cram_first.mix(ch_tumor_cram).groupTuple()
+    // Add intervals back
+    ch_pt_intervals = input_intervals.map{meta, n_cram, n_crai, t_cram, t_crai, intervals -> [[id: meta.patient, patient: meta.patient, normal_id: meta.normal_id, sex: meta.sex, num_intervals: meta.num_intervals], intervals]}.unique().groupTuple()
```
the meta mapping can be done with addition/subMap, which is safer then manually setting all keys, something we learned recently and have updated in all other places

FriederikeHanssen commented on subworkflows/local/bam_variant_calling_somatic_mutect2_ms/main.nf:
```diff
@@ -0,0 +1,324 @@
+    // Contamination and segmentation tables created using calculatecontamination on the pileup summary table.
+    // Do some channel magic to generate tumor-normal pairs again.
+    // This is necessary because we generated one normal pileup summary for each patient but we need run calculate contamination for each tumor-normal pair.
+    ch_tumor_pileup_tables = gather_table_tumor.map{meta, table -> [[id: meta.patient, patient: meta.patient, normal_id: meta.normal_id, sex: meta.sex, num_intervals: meta.num_intervals], meta.id, table]}
+    ch_normal_pileup_tables = gather_table_normal.map{meta, table -> [[id: meta.patient, patient: meta.patient, normal_id: meta.normal_id, sex: meta.sex, num_intervals: meta.num_intervals], meta.id, table]}
+    ch_calculatecontamination_in_tables = ch_tumor_pileup_tables.combine(ch_normal_pileup_tables, by:0).map{meta, tumor_id, tumor_table, normal_id, normal_table ->
+                                                                                                            new_meta = [ id: tumor_id + "_vs_" + normal_id, patient: meta.patient, normal_id: meta.normal_id, sex: meta.sex, num_intervals: meta.num_intervals]
```
all the meta maps should be manipulated with addition or subMap for safety

FriederikeHanssen reviewed (commented):
- Can we simplify the configs a bit to have the parts that are the same for single & multi sharedOr would that complicate things?
- not all publishing is done to the same folder (single/multi) (like GetPileupsummaries), would it make sense to have one folder for each case, or just in general one mutect2 output folder, since both modes can't be mixed
- from what I can see the subworkflows are pretty much the same starting at line 63 (somatic here, but simliar for tumor-only). Can we extract the input channel magic from both (multi-sample/single) and then call the mutect2 subworkflow for both, then only one place needs to be maintained & easier to xtract it to nf-core subworkflows at some point

FriederikeHanssen commented:
Amazing @berguner ! Thank you so much for adding this 🚀

berguner commented on subworkflows/local/bam_variant_calling_somatic_mutect2_ms/main.nf:
```diff
@@ -0,0 +1,324 @@
+
+    // Prepare multi sample Mutect2 variant calling channels
+    // Separate normal cram files
+    ch_normal_cram = input_intervals.map{meta, n_cram, n_crai, t_cram, t_crai, intervals -> [[id: meta.patient, patient: meta.patient, normal_id: meta.normal_id, sex: meta.sex, num_intervals: meta.num_intervals], n_cram, n_crai]}.unique().groupTuple()
```
You are right, this was because I branched from master and the input channel had intervals mixed in. We shouldn't need `unique` if we start with the cram inputs only but that needs some tweaking.

berguner replied:
Regarding the `size` parameter for `groupTuple`, it seems like there isn't an easy way to get the number of cram files per patient at the subworkflow level. This bottleneck happens upstream of the subwf [here](https://github.com/nf-core/sarek/blob/ed1cc8499366dcefea216fe37e36c6189537d57b/workflows/sarek.nf#L872) and I don't know how to overcome that.

berguner commented:
Thanks @maxulysse and @FriederikeHanssen for the code reviews! As you can tell I am not very experienced with Nextflow development, this was my first (sub)workflow implementation. I don't want to be in the way when it comes to best NF coding practices and would very much appreciate learning from your reviews. So, feel free to commit any improvements that you see fit as long as the functionality is reserved.

The reason why I tried to separate the multi sample subworkflows as much as possible was to avoid merge conflicts and preserve backwards compatibility for the single tumor sample mode. In principle, multi sample mode could replace the single sample mode because it works with single tumor also but the VCF file naming has to change. I am planning to switch to multi sample completely when this is ready but others might want to continue with the `_vs_` VCF files.

berguner commented on conf/modules/mutect2.config:
```diff
@@ -14,110 +14,197 @@
+            ext.prefix       = { meta.num_intervals <= 1 ? "${meta.id}.mutect2" : "${meta.id}.mutect2.${intervals.simpleName}" }
+            ext.when         = { params.tools && params.tools.split(',').contains('mutect2') && params.mutect2_multi_sample }
+            ext.args         = { params.ignore_soft_clipped_bases ? "--dont-use-soft-clipped-bases true --f1r2-tar-gz ${task.ext.prefix}.f1r2.tar.gz" : "--f1r2-tar-gz ${task.ext.prefix}.f1r2.tar.gz" }
+            publishDir       = [
+                mode: params.publish_dir_mode,
+                path: { "${params.outdir}/variant_calling/" },
+                pattern: "*{vcf.gz,vcf.gz.tbi,stats}",
+                saveAs: { meta.num_intervals > 1 ? null : "mutect2/multi_sample/${meta.id}/${it}" }
+            ]
```
I think we should keep the `multi_sample` subfolder to emphasise the distinction of the new naming convention which is based on patient ID.

berguner replied:
After some head scratching, I also decided to get rid of `multi_sample` subfolder. It's cleaner and we can have more coherent configuration for single sample and multi sample modes.

berguner commented on subworkflows/local/bam_variant_calling_somatic_all/main.nf:
```diff
@@ -10,6 +10,8 @@ include { BAM_VARIANT_CALLING_SOMATIC_ASCAT             } from '../bam_variant_c
 include { BAM_VARIANT_CALLING_SOMATIC_MANTA             } from '../bam_variant_calling_somatic_manta/main'
 include { BAM_VARIANT_CALLING_SOMATIC_MUTECT2           } from '../bam_variant_calling_somatic_mutect2/main'
+// Variant calling for patients with multiple tumor samples 
+include { BAM_VARIANT_CALLING_SOMATIC_MUTECT2_MS        } from '../bam_variant_calling_somatic_mutect2_ms/main'
```
Done!

berguner commented:
Hi @maxulysse , @FriederikeHanssen ,
I think I addressed almost all of the comments with the latest commit. I incorporated the multi sample mode into the existing Mutect2 subworkflows and removed the `_ms` ones. I also opted for removing the `multi_sample` output subfolder and cleaned up `mutect2.config`.

The only unsolved issue is using `groupTuple` without size parameter when grouping some output channels patient-wise. One solution that I can think of is capturing the number of tumor samples per patient upstream in the pipeline and feeding it to the mutect2 subworkflows. I am reluctant to attempt this now because of time constraints. Of course, it would be great if you can think of a simpler solution.

Please have a look at the current state of the code and let me know what you think.

maxulysse commented on subworkflows/local/bam_variant_calling_somatic_all/main.nf:
```diff
@@ -2,17 +2,17 @@
-include { BAM_VARIANT_CALLING_SOMATIC_STRELKA           } from '../bam_variant_calling_somatic_strelka/main'
-include { BAM_VARIANT_CALLING_SOMATIC_TIDDIT            } from '../bam_variant_calling_somatic_tiddit/main'
-include { MSISENSORPRO_MSI_SOMATIC                      } from '../../../modules/nf-core/msisensorpro/msi_somatic/main'
+include { BAM_VARIANT_CALLING_CNVKIT                              } from '../bam_variant_calling_cnvkit/main'
+include { BAM_VARIANT_CALLING_FREEBAYES                           } from '../bam_variant_calling_freebayes/main'
+include { BAM_VARIANT_CALLING_MPILEUP as MPILEUP_NORMAL           } from '../bam_variant_calling_mpileup/main'
+include { BAM_VARIANT_CALLING_MPILEUP as MPILEUP_TUMOR            } from '../bam_variant_calling_mpileup/main'
+include { BAM_VARIANT_CALLING_SOMATIC_ASCAT                       } from '../bam_variant_calling_somatic_ascat/main'
+include { BAM_VARIANT_CALLING_SOMATIC_CONTROLFREEC                } from '../bam_variant_calling_somatic_controlfreec/main'
+include { BAM_VARIANT_CALLING_SOMATIC_MANTA                       } from '../bam_variant_calling_somatic_manta/main'
+include { BAM_VARIANT_CALLING_SOMATIC_MUTECT2                     } from '../bam_variant_calling_somatic_mutect2/main'
+include { BAM_VARIANT_CALLING_SOMATIC_STRELKA                     } from '../bam_variant_calling_somatic_strelka/main'
+include { BAM_VARIANT_CALLING_SOMATIC_TIDDIT                      } from '../bam_variant_calling_somatic_tiddit/main'
+include { MSISENSORPRO_MSI_SOMATIC                                } from '../../../modules/nf-core/msisensorpro/msi_somatic/main'
```
```suggestion
include { BAM_VARIANT_CALLING_CNVKIT                    } from '../bam_variant_calling_cnvkit/main'
include { BAM_VARIANT_CALLING_FREEBAYES                 } from '../bam_variant_calling_freebayes/main'
include { BAM_VARIANT_CALLING_MPILEUP as MPILEUP_NORMAL } from '../bam_variant_calling_mpileup/main'
include { BAM_VARIANT_CALLING_MPILEUP as MPILEUP_TUMOR  } from '../bam_variant_calling_mpileup/main'
include { BAM_VARIANT_CALLING_SOMATIC_ASCAT             } from '../bam_variant_calling_somatic_ascat/main'
include { BAM_VARIANT_CALLING_SOMATIC_CONTROLFREEC      } from '../bam_variant_calling_somatic_controlfreec/main'
include { BAM_VARIANT_CALLING_SOMATIC_MANTA             } from '../bam_variant_calling_somatic_manta/main'
include { BAM_VARIANT_CALLING_SOMATIC_MUTECT2           } from '../bam_variant_calling_somatic_mutect2/main'
include { BAM_VARIANT_CALLING_SOMATIC_STRELKA           } from '../bam_variant_calling_somatic_strelka/main'
include { BAM_VARIANT_CALLING_SOMATIC_TIDDIT            } from '../bam_variant_calling_somatic_tiddit/main'
include { MSISENSORPRO_MSI_SOMATIC                      } from '../../../modules/nf-core/msisensorpro/msi_somatic/main'
```

maxulysse commented on subworkflows/local/bam_variant_calling_somatic_all/main.nf:
```diff
@@ -178,26 +178,23 @@ workflow BAM_VARIANT_CALLING_SOMATIC_ALL {
         out_msisensorpro = out_msisensorpro.mix(MSISENSORPRO_MSI_SOMATIC.out.output_report)
     }
 
-    // MUTECT2
-    if (tools.split(',').contains('mutect2')) {
```
I'd like to keep that very much please

berguner replied:
Of course we need it, I put it back now.

maxulysse commented on subworkflows/local/bam_variant_calling_tumor_only_all/main.nf:
```diff
@@ -3,14 +3,14 @@
-include { BAM_VARIANT_CALLING_TUMOR_ONLY_CONTROLFREEC } from '../bam_variant_calling_tumor_only_controlfreec/main'
-include { BAM_VARIANT_CALLING_TUMOR_ONLY_MANTA        } from '../bam_variant_calling_tumor_only_manta/main'
-include { BAM_VARIANT_CALLING_TUMOR_ONLY_MUTECT2      } from '../bam_variant_calling_tumor_only_mutect2/main'
+include { BAM_VARIANT_CALLING_CNVKIT                            } from '../bam_variant_calling_cnvkit/main'
+include { BAM_VARIANT_CALLING_FREEBAYES                         } from '../bam_variant_calling_freebayes/main'
+include { BAM_VARIANT_CALLING_MPILEUP                           } from '../bam_variant_calling_mpileup/main'
+include { BAM_VARIANT_CALLING_SINGLE_STRELKA                    } from '../bam_variant_calling_single_strelka/main'
+include { BAM_VARIANT_CALLING_SINGLE_TIDDIT                     } from '../bam_variant_calling_single_tiddit/main'
+include { BAM_VARIANT_CALLING_TUMOR_ONLY_CONTROLFREEC           } from '../bam_variant_calling_tumor_only_controlfreec/main'
+include { BAM_VARIANT_CALLING_TUMOR_ONLY_MANTA                  } from '../bam_variant_calling_tumor_only_manta/main'
+include { BAM_VARIANT_CALLING_TUMOR_ONLY_MUTECT2                } from '../bam_variant_calling_tumor_only_mutect2/main'
```
```suggestion
include { BAM_VARIANT_CALLING_CNVKIT                  } from '../bam_variant_calling_cnvkit/main'
include { BAM_VARIANT_CALLING_FREEBAYES               } from '../bam_variant_calling_freebayes/main'
include { BAM_VARIANT_CALLING_MPILEUP                 } from '../bam_variant_calling_mpileup/main'
include { BAM_VARIANT_CALLING_SINGLE_STRELKA          } from '../bam_variant_calling_single_strelka/main'
include { BAM_VARIANT_CALLING_SINGLE_TIDDIT           } from '../bam_variant_calling_single_tiddit/main'
include { BAM_VARIANT_CALLING_TUMOR_ONLY_CONTROLFREEC } from '../bam_variant_calling_tumor_only_controlfreec/main'
include { BAM_VARIANT_CALLING_TUMOR_ONLY_MANTA        } from '../bam_variant_calling_tumor_only_manta/main'
include { BAM_VARIANT_CALLING_TUMOR_ONLY_MUTECT2      } from '../bam_variant_calling_tumor_only_mutect2/main'
```

maxulysse commented on subworkflows/local/bam_variant_calling_tumor_only_all/main.nf:
```diff
@@ -106,25 +106,23 @@ workflow BAM_VARIANT_CALLING_TUMOR_ONLY_ALL {
         versions = versions.mix(BAM_VARIANT_CALLING_FREEBAYES.out.versions)
     }
 
-    // MUTECT2
-    if (tools.split(',').contains('mutect2')) {
```
That would be nice to keep please

berguner replied:
Done!

maxulysse commented on subworkflows/local/bam_variant_calling_somatic_all/main.nf:
```diff
@@ -194,11 +194,11 @@ workflow BAM_VARIANT_CALLING_SOMATIC_ALL {
-        versions = versions.mix(BAM_VARIANT_CALLING_SOMATIC_MUTECT2.out.versions)
     }
 
+    vcf_mutect2 = BAM_VARIANT_CALLING_SOMATIC_MUTECT2.out.vcf_filtered
+    versions = versions.mix(BAM_VARIANT_CALLING_SOMATIC_MUTECT2.out.versions)
+
```
That should be in the if block

berguner replied:
ahh, yes of course 😅

berguner replied:
done!

maxulysse commented on subworkflows/local/bam_variant_calling_tumor_only_all/main.nf:
```diff
@@ -121,11 +121,11 @@ workflow BAM_VARIANT_CALLING_TUMOR_ONLY_ALL {
-        versions = versions.mix(BAM_VARIANT_CALLING_TUMOR_ONLY_MUTECT2.out.versions)
     }
 
+    vcf_mutect2 = BAM_VARIANT_CALLING_TUMOR_ONLY_MUTECT2.out.vcf_filtered
+    versions = versions.mix(BAM_VARIANT_CALLING_TUMOR_ONLY_MUTECT2.out.versions)
+
```
same here

berguner replied:
Done!

maxulysse commented on CHANGELOG.md:
```diff
@@ -96,7 +97,7 @@ Bierikjávrre is one of the largest lake in Sarek.
 - [#981](https://github.com/nf-core/sarek/pull/981) - Added documentation on generating ASCAT resources for exome and targeted sequencing
 - [#1041](https://github.com/nf-core/sarek/pull/1041) - Add params `vep_custom_args` to let user specify custom params more easily for `VEP`
-- [#1045](https://github.com/nf-core/sarek/pull/1045) - Add `public_aws_ecr` for using ECR hosted containers
+- [#1045](https://github.com/nf-core/sarek/pull/1045) - Add `public_aws_ecr` for using ECR hosted containers.
```
```suggestion
- [#1045](https://github.com/nf-core/sarek/pull/1045) - Add `public_aws_ecr` for using ECR hosted containers
```

maxulysse commented on nextflow.config:
```diff
@@ -52,6 +52,7 @@ params {
 
     // Variant Calling
     only_paired_variant_calling = false // if true, skips germline variant calling for normal-paired samples
+    joint_mutect2               = false // if true, enables patient-wise multi-sample somatic variant calling
```
Can it be put after the joint_germline params ?

berguner replied:
Done!

maxulysse commented:
Looks amazing.
We're now missing just some documentation about it

FriederikeHanssen commented on conf/modules/mutect2.config:
```diff
@@ -14,110 +14,137 @@
+                    pattern: "*{vcf.gz,vcf.gz.tbi,stats}",
+                    saveAs: { meta.num_intervals > 1 ? null : "mutect2/${meta.id}/${it}" }
+                ]
+            }
```
```suggestion
        }
```

FriederikeHanssen commented on conf/modules/mutect2.config:
```diff
@@ -14,110 +14,137 @@
-        ]
-    }
+        withName: 'GETPILEUPSUMMARIES_.*' {
+            ext.prefix        = { meta.num_intervals <= 1 ? "${meta.id}.mutect2" : "${meta.id}.mutect2.${intervals.simpleName}" }
```
I think this line can be removed, already taken care of in the more general one above
```suggestion
```

FriederikeHanssen commented on conf/modules/mutect2.config:
```diff
@@ -14,110 +14,137 @@
+        if (params.joint_mutect2) {
+            // JOINT (MULTI SAMPLE) MUTECT2
+            withName: 'GETPILEUPSUMMARIES.*' {
+                ext.prefix        = { meta.num_intervals <= 1 ? "${meta.id}.mutect2" : "${meta.id}.mutect2.${intervals.simpleName}" }
```
this one as well, you just want to overwrite the publishing directory, right?
```suggestion
```

FriederikeHanssen commented on conf/modules/mutect2.config:
```diff
@@ -14,110 +14,137 @@
+            }
+
+            withName: 'GATHERPILEUPSUMMARIES.*' {
+                ext.prefix       = { "${meta.id}.mutect2" }
```
```suggestion
```

FriederikeHanssen commented on conf/modules/mutect2.config:
```diff
@@ -14,110 +14,137 @@
+                    saveAs: { filename -> filename.equals('versions.yml') ? null : filename }
+                ]
+            }
+                withName: 'CALCULATECONTAMINATION' {
```
```suggestion
           withName: 'CALCULATECONTAMINATION' {
```

FriederikeHanssen commented on conf/modules/prepare_genome.config:
```diff
@@ -76,7 +76,7 @@ process {
 
     withName: 'TABIX_DBSNP' {
-        ext.when         = { !params.dbsnp_tbi && params.dbsnp && ((params.step == "mapping" || params.step == "markduplicates" || params.step == "prepare_recalibration") || params.tools && (params.tools.split(',').contains('controlfreec') || params.tools.split(',').contains('haplotypecaller') || params.tools.split(',').contains('mutect2'))) }
+        ext.when         = { !params.dbsnp_tbi && params.dbsnp && (params.step == "mapping" || params.step == "markduplicates" || params.step == "prepare_recalibration") || params.tools && ( params.tools.split(',').contains('controlfreec') || params.tools.split(',').contains('haplotypecaller') || params.tools.split(',').contains('mutect2')) }
```
why are you removing the `(` here? It changes the expression I think:


0 && 1 when dbsnp doesn't exists and dbsnp does
AND ( (when starting from mapping/md/prepare_recal)  OR Controlfreec OR Haplytypecalle OR Mutect )

!= AND (when starting from mapping/md/prepare_recal)  OR Controlfreec OR Haplytypecalle OR Mutect

berguner replied:
Good catch, I think I deleted them by mistake when I removed the now obsolete `_MS` subwfs.

FriederikeHanssen commented on subworkflows/local/bam_variant_calling_somatic_mutect2/main.nf:
```diff
@@ -36,8 +37,23 @@ workflow BAM_VARIANT_CALLING_SOMATIC_MUTECT2 {
+        // Extract tumor cram files
+        ch_tumor_cram = input.map{ meta, cram, crai -> [ meta - meta.subMap('tumor_id') + [id:meta.patient], cram[1], crai[1] ] }
+        // Merge normal and tumor crams by patient
+        ch_tn_cram = ch_normal_cram.mix(ch_tumor_cram).groupTuple()
```
oh no a groupTuple without a size. This will stall the workflow and cause a bottleneck. I don't have a good idea at the moment on how to properly retrieve the size. Can you open an issue please, that we need to figure this out and fix at some point?

berguner replied:
https://github.com/nf-core/sarek/issues/1141

FriederikeHanssen commented on subworkflows/local/bam_variant_calling_somatic_mutect2/main.nf:
```diff
@@ -24,6 +24,7 @@ workflow BAM_VARIANT_CALLING_SOMATIC_MUTECT2 {
     panel_of_normals          // channel: /path/to/panel/of/normals
     panel_of_normals_tbi      // channel: /path/to/panel/of/normals/index
     intervals                 // channel: [mandatory] [ intervals, num_intervals ] or [ [], 0 ] if no intervals
+    joint_mutect2                 // boolean: [mandatory] [default: false] run mutect2 in joint mode
```
```suggestion
    joint_mutect2            // boolean: [mandatory] [default: false] run mutect2 in joint mode
```

FriederikeHanssen commented on subworkflows/local/bam_variant_calling_somatic_mutect2/main.nf:
```diff
@@ -89,7 +105,15 @@ workflow BAM_VARIANT_CALLING_SOMATIC_MUTECT2 {
+    if (joint_mutect2) {
+        // Prepare input channel for normal pileup summaries.
+        // Remember, the input channel contains tumor-normal pairs, so there will be multiple copies of the normal sample for each tumor for a given patient.
+        // Therefore, we use unique function to generate normal pileup summaries once for each patient for better efficiency.
```
this should actually also be the case without joint calling. so i think this can be the general setup. Good eye

berguner replied:
To do this we need to remove `tumor_id` but that breaks publishDir path for normal `pileups.table`, for example it becomes `variant_calling/mutect2/null_vs_sample3/` with test data.

berguner replied:
Just pushed https://github.com/nf-core/sarek/pull/1013/commits/b3fec41c955d4abc34e8798866980b6c0099cdf4 which should avoid the redundant normal pileup jobs. `pileups.table` files will be saved under `mutect2/sample/` subfolders.

FriederikeHanssen commented on subworkflows/local/bam_variant_calling_somatic_mutect2/main.nf:
```diff
@@ -118,23 +142,45 @@ workflow BAM_VARIANT_CALLING_SOMATIC_MUTECT2 {
+    if (joint_mutect2) {
+        // Do some channel magic to generate tumor-normal pairs again.
+        // This is necessary because we generated one normal pileup summary for each patient but we need run calculate contamination for each tumor-normal pair.
+        pileup_table_tumor = Channel.empty().mix(GATHERPILEUPSUMMARIES_TUMOR.out.table, pileup_table_tumor_branch.no_intervals).map{meta, table -> [ meta - meta.subMap('normal_id', 'tumor_id', 'num_intervals') + [id:meta.patient], meta.id, table ] }
+        pileup_table_normal= Channel.empty().mix(GATHERPILEUPSUMMARIES_NORMAL.out.table, pileup_table_normal_branch.no_intervals).map{meta, table -> [ meta - meta.subMap('normal_id', 'tumor_id', 'num_intervals') + [id:meta.patient], meta.id, table ] }
+        ch_calculatecontamination_in_tables = pileup_table_tumor.combine(
+            pileup_table_normal, by:0).map{
```
same here, I think this can be done regardless of scenario

berguner replied:
see https://github.com/nf-core/sarek/pull/1013#discussion_r1252161708

FriederikeHanssen reviewed (changes requested):
Fantastic work! I think you also found two inefficient computations we do, that can be fixed with your code

berguner commented on subworkflows/local/bam_variant_calling_somatic_all/main.nf:
```diff
@@ -50,6 +52,7 @@ workflow BAM_VARIANT_CALLING_SOMATIC_ALL {
     vcf_strelka         = Channel.empty()
     out_msisensorpro    = Channel.empty()
     vcf_mutect2         = Channel.empty()
+    vcf_mutect2_ms      = Channel.empty()
```
You were right, they are mixed now.

maxulysse reviewed (approved):
AMAZING
